src/choker.cpp

```diff
@@ -59,6 +59,14 @@ namespace {
 		return 0;
 	}
 
+	// return true if peer is suspected of dishonesty
+	bool dishonest_peer(peer_connection const* peer, int piece_length)
+	{
+		// some clients (e.g. XunLei) always claim they have zero pieces as fresh starter,
+		// no matter how many we have uploaded.
+		return peer->num_have_pieces() == 0 && peer->uploaded_at_last_round() >= piece_length;
```

Comment:
I think you'll need to add some factor to the piece length here. I can think of three considerations:

1. The blocks we've uploaded may not have been part of the same piece, and the peer may still be stuck waiting for the last stragglers to complete those pieces.
2. There will be some delay between the peer completing the piece and us receiving its "HAVE" message. It will have to check its hash, maybe write it to disk and send back the message over the wire. There will be *at least* one round-trip time of delay, but probably more.
3. The peer may chose to batch HAVE messages it sends back, which will contribute to the latency we experience from uploading and receiving the HAVE notification.

Another consideration is that this heuristic is probably too simplistic. All it would take would be for xunlei (or whichever client is trying to cheat) would be to always report having 1 piece.

I think more accurate way of defeating this would be to remember how much we've uploaded to a peer and use the max of that and `num_have_pieces()` in this anti-leech choker.

I believe we already store the total amount of payload we've uploaded to a peer, and we preserve it between reconnects even.

Reply:
1. Correct me if wrong, we prefer uploading blocks in the same piece, unless part of it has gotten from others. In that situation, peer would have at least 1 piece eventually.
2. Yes, it may take several rounds to comfirm (by trip delay and batching messages). But they will get nearly hightest score in next round, I think it's worth if we choke those pure leechers always higher than them. Even there is no cheater, innocently choked for several round-time won't hurt too much.

Anyway, I'm OK with an extra factor. @aresch , how many would you think is appropriate?

3. XunLei is the largest download company in China, which has it' own download network and cache servers. It won't share any, the high score by cheeting is merely a side-effect. I don't think it would likely change the cheating means.
If situation change, we can improve `dishonest_peer()` or even the algorithm in the future, for now it looks good enough to me.
4. @ssiloti , no algorithm is perfert, people choose this anti-leech choker because they want anti-leech, and this patch fixes a most obvious cheating loophole. For those believing other unchoke algorithm, it won't hurt anyway.